### PR TITLE
Support versioned key

### PR DESCRIPTION
**Pull request by chenliu0831**
The bodyfitter worker will be expected s3 urls with versionId metadata now.

The context is the effort to remove our original file upload->copy to storage process in the server code which adds complexity.

### CODE REVIEW

**Comment by qmaurmann** [comment]
Right, most future bodylabs API's (e.g. 2k & shape) will depart from RED v1 in using versioned paths for easy immutability (rather than a versioned upload path, copied once to an unversioned storage path). So this either has to be supported in `baiji` (which would be my preference), or applications using versioned paths will have to abandon `baiji` in favor of directly using `boto`.

I think @dlsmith should weigh in on this too, as I remember him being skeptical of my suggestion in May/June to add versioning to paths (though my suggestion was slightly different-- to add `version` as a query param to the bodylabs S3 URL format).

I also think that if `version` is added to `baiji`, then we should carefully make sure it is supported the same in all the relevant operations (not just `cp`, but also `exists`, `cat`, and possibly others).

**Comment by chenliu0831** [comment]
yeah I agree. If versioning support in `baiji` makes our brain twisted too much, I think it's fairly easy to make a helper util in the runner code for native versioned S3 keys (in boto).

**Comment by algrs** [comment]
I think it's sane to support versioning here in baiji. I'd like to support it everywhere, as @qmaurmann suggests. Probably also need either an option on `ls` or some sort of list-versions command. 

A while back we switched from a fairly clunky local backup of s3 to activating versioning on all buckets, and since then I've had in the back of my mind the need for a way to work with versions...

File: baiji/connection.py [hunk 1]
```diff
@@ -66,7 +66,7 @@ def _lookup(self, bucket_name, key, cache_buckets=None):
 
         return bucket.lookup(key)
 
-    def cp(self, key_or_file_from, key_or_file_to, force=False, progress=False, policy=None, preserve_acl=False, encoding=None, encrypt=True, gzip=False, content_type=None, guess_content_type=False, metadata=None, skip=False, validate=True, max_size=None):
+    def cp(self, key_or_file_from, key_or_file_to, force=False, progress=False, policy=None, preserve_acl=False, encoding=None, encrypt=True, gzip=False, content_type=None, guess_content_type=False, metadata=None, skip=False, validate=True, max_size=None, version=None):
```
Comment (algrs) [inline baiji/connection.py]: You're using both `version` and `version_id` throughout... Is there a reason why you chose `version` in some places and `version_id` in others?

File: baiji/copy.py [hunk 2]
```diff
@@ -51,19 +51,35 @@ def etag(self):
...
+            from boto.exception import S3ResponseError
+            from baiji.exceptions import InvalidVersionID
+
             if self.is_file:
                 raise ValueError("S3CopyOperation.CopyableKey.lookup called for local file")
-            key = self.bucket.lookup(self.remote_path)
+
+            key = None
+
+            try:
```
Comment (algrs) [inline baiji/copy.py]: This is functionality that I expect we'll wind up needing in many places as we add support for versions; probably worth wrapping it in a utility function.

File: baiji/copy.py [hunk 3]
```diff
@@ -333,7 +359,7 @@ def upload_multipart(self):
     def upload_direct(self):
         import math
         from baiji.util.with_progressbar import FileTransferProgressbar
-        key = self.dst.create()
+        key = self.dst.create(version_id=self.version)
```
Comment (algrs) [inline baiji/copy.py]: I don't think this makes sense -- if you're uploading, it should make a new version, so you should only be able to specify the version for the source of the copy. Versions are immutable, enforced by s3, so if you try to write to a specific version I'd expect this to break... Can you test this?

Comment (chenliu0831) [inline baiji/copy.py]: ah yes, I'll test and remove it. The API makes it look like you can use your own versionID, but probably not.

File: baiji/connection.py (hunk 1 above)
Comment (chenliu0831) [inline baiji/connection.py]: ok I'll conform to `version_id`

**Comment by dlsmith** [comment]
Full support in baiji sounds good to me.

**Comment by chenliu0831** [comment]
IMO the following commands should be supporting versions (against remote s3 keys):

- cp: s3 to s3
- sync
- exists
- rm
- get_url
- metadata commands: size, etag, md5, exists

Most interface has to be extended with an optional `version_id` arg.

For `ls` and `glob` it seems best if we invent some URL format for the version id, but I feel it'll affect the current path parse logic a lot, probably best to be implemented separately.

**Comment by chenliu0831** [comment]
@algrs added cli flags. I also added support on `exists`, `mv`, `rm`, and `list-versions` as an `ls` option, which looks like:

```

$ s3 ls -l s3://baiji-test-versioned --list-versions

43.0b   2017-04-06 16:54:19 enc s3://baiji-test-versioned/FOO/A_preexisting_file.in F5EV24LVHe3P0ejKkgASHuYvsOAnFZgR
43.0b   2017-04-06 17:09:05 enc s3://baiji-test-versioned/FOO/test.in   .6dOQmqDWTOhAiiwtJU_gn1tzZOVfRYq
43.0b   2017-04-06 17:09:05 enc s3://baiji-test-versioned/FOO/test.in   .6dOQmqDWTOhAiiwtJU_gn1tzZOVfRYq
```

**Comment by chenliu0831** [comment]
Thanks @algrs. Could you add `baiji-test-versioned` bucket access to the test AWS key circleCI is using? Thanks

**Comment by dlsmith** [comment]
@chenliu0831 I removed myself as a reviewer, as I trust that you and Alex covered this well. Let me know if you guys would like another set of eyes on this though.

**Comment by algrs** [comment]
@chenliu0831 Permissions added

**Comment by chenliu0831** [comment]
@algrs Seems write permission was not added? It has a `mv` test which delete the files

**Comment by algrs** [comment]
It's a full read/write permission.

**Comment by chenliu0831** [comment]
hmm It seems still having trouble delete a versioned key (`got 403` from `s3://baiji-test-versioned/FOO/A_preexisting_file.md`). Could you double check the permission is added to all the subdirectory? Read seems fine.

**Comment by algrs** [comment]
It's the standard read/write permission that we use everywhere... Are you trying to delete a key or permanently expunge a version?

**Comment by chenliu0831** [comment]
@algrs hmm I think currently the `mv` when provide with a versionId looks more like expunging a version. Is that separate from write permission?

**Comment by algrs** [comment]
Hmm.... I'll admit, I don't like the idea that `mv` would expunge, but I see how that's the obvious interpretation. I think the better thing is to have `mv a b --version_id v` give an error message saying that "To avoid the risk of accidentally expunging backups, please copy the file and then explicitly expunge the version if that's your intent"

**Comment by algrs** [comment]
Yeah, the more I think about it, the more it seems that the semantics of moving a version is needlessly risky, especially given that move is basically a convenience method for copy+delete. Sorry for asking you to add it -- I was wrong to think it was a good idea.

**Comment by chenliu0831** [comment]
Don't worry I'll remove it. It does seem confusing semantically.

**Comment by qmaurmann** [comment]
Yeah, I am also leaning on the principle that if Alex is happy (with interface changes to `baiji`), then I am happy. (Unless a second reviewer is specifically required.)

**Comment by algrs** [comment]
Alex is happy. 👍

**Pull request merged**